File: algorithms/dijkstra.py
def init(len):
    mark = [0]*len
    dist = [999999]*len
    return mark, dist


def min(mark, dist, length):
    min = 999999
    for j in range(length):
        if mark[j] == 0 and dist[j] < min:
            min = dist[j]
            k = j
    return k



def dijkstra(a, src, dist, mark):
    dist[src] = 0;
    for i in a:
        min_idx = min(mark, dist, len(a))
        mark[min_idx] = 1
        for u in range(len(a)):
            if a[min_idx][u] != 0 and mark[u] == 0 and dist[min_idx] + a[min_idx][u] < dist[u]:
                dist[u] = dist[min_idx] + a[min_idx][u]

def main():
    a = [
        [0, 6, 0, 1, 0],
        [6, 0, 5, 2, 2],
        [0, 5, 0, 0, 5],
        [1, 2, 0, 0, 1],
        [0, 2, 5, 1, 0]
    ]
    mark, dist = init(len(a))
    dijkstra(a, 0, dist, mark)
    print (dist)

main()

File: algorithms/dijkstra.c
#include "stdio.h"
#define V 5
#define infinity 9999999

int mark[V];
int dist[V];
int path[V];
void dijkstra(int a[][V], int start)
{
    for (int i = 0; i < V; i++)
    {
        // *(dist + i) = infinity;
        dist[i] = infinity;
        // *(mark + i) = 0;
        mark[i] = 0;
    }
    *(dist + start) = 0;
    for (int i=0; i<V; i++) {
        int min = infinity;
        int k;
        for (int j=0; j<V; j++) {
            if (mark[j] == 0 && dist[j] < min) {
                min = dist[j];
                k = j;
            }
        }
        mark[k] = 1;
        for (int u=0; u<V; u++){
            if (mark[u] == 0 && a[k][u] != 0 && dist[k] + a[k][u] < dist[u]) {
                dist[u] = dist[k] + a[k][u];
            }
        }
    }
}
int main()
{
    int graph[V][V] = {{0, 6, 0, 1, 0},
                        {6, 0, 5, 2, 2},
                        {0, 5, 0, 0, 5},
                        {1, 2, 0, 0, 1},
                        {0, 2, 5, 1, 0},
                       };

    printf ("Graph: \n");
    int *pa = (int*)graph;
    for (int i = 0; i < V; i++)
    {
        for (int j = 0; j < V; j++)
        {
            // printf("%d ", *(*(graph + i) + j));
            printf ("%d ", *(pa + i*V + j));
        }
        printf("\n");
    }
    dijkstra(graph, 0);
    printf ("Shortest path from src to other nodes: \n");
    for (int i=0; i<V; i++) {
        printf ("%d ", dist[i]);
    }
}